### PR TITLE
Enable custom profile pictures

### PR DESCRIPTION
**Pull request by msgmaxim**
Screen Shot 2019-11-11 at 3 12 14 pm

- Redesigned UI for updating display name (now "Edit Profile") as a React component adding the ability to upload a profile image.
- Added "avatar pointers" to both DM and Public Chat messages, parsing them on the receiving side to update conversation profiles.
- Additionally fixed display names not being updated on receiving Public Chat messages.
- Addresses oxen-io/session-desktop-temp#1097 (does not close, as we need mobile changes as well)

### CODE REVIEW

File: js/modules/loki_app_dot_net_api.js [hunk 1]
```diff
@@ -978,19 +989,27 @@ class LokiPublicChannelAPI {
       LokiPublicChannelAPI.getAnnotationFromPreview
     );
 
+    const avatarAnnotation = window.serializeAttachmentPointer(
+      data.profile.avatar
+    );
+
+    // console.error("avatar annotation:", avatarAnnotation);
```
Comment (BeaudanBrown) [inline js/modules/loki_app_dot_net_api.js]: Old

File: libtextsecure/message_receiver.js [hunk 2]
```diff
@@ -154,7 +155,16 @@ MessageReceiver.prototype.extend({
     };
     this.httpPollingResource.handleMessage(message, options);
   },
-  handleUnencryptedMessage({ message }) {
+  async handleUnencryptedMessage({ message }) {
+    const conversation = await window.ConversationController.getOrCreateAndWait(
```
Comment (BeaudanBrown) [inline libtextsecure/message_receiver.js]: Move the conversation get inside the if below?

File: libtextsecure/message_receiver.js [hunk 3]
```diff
@@ -1048,6 +1058,32 @@ MessageReceiver.prototype.extend({
...
+    const newProfile = conversation.get('profile') || {};
+
+    newProfile.displayName = profile.displayName;
+
+    // TODO: may need to allow users to reset their avatars to null
+    if (profile.avatar) {
+      const prevPointer = conversation.get('avatarPointer');
+      const needsUpdate =
+        !prevPointer ||
+        (prevPointer && !_.isEqual(prevPointer, profile.avatar));
```
Comment (BeaudanBrown) [inline libtextsecure/message_receiver.js]: Left half of this && is implied by the ||

File: ts/components/EditProfileDialog.tsx [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+      return;
+    }
+
+    const avatar =
+      this.inputEl &&
+      this.inputEl.current &&
+      this.inputEl.current.files &&
+      this.inputEl.current.files.length > 0
+        ? this.inputEl.current.files[0]
+        : null;
```
Comment (BeaudanBrown) [inline ts/components/EditProfileDialog.tsx]: Thank you javascript :^)

File: libtextsecure/message_receiver.js (hunk 3 above)
Comment (msgmaxim) [reply libtextsecure/message_receiver.js]: updated

File: libtextsecure/message_receiver.js (hunk 2 above)
Comment (msgmaxim) [reply libtextsecure/message_receiver.js]: updated

File: js/modules/loki_app_dot_net_api.js (hunk 1 above)
Comment (msgmaxim) [reply js/modules/loki_app_dot_net_api.js]: updated

**Comment by msgmaxim** [comment]
Updated the PR with a minor change in the protocol and fixing conflicts with *clearnet*.

File: js/background.js [hunk 5]
```diff
@@ -832,15 +832,70 @@
         ourNumber,
         'private'
       );
+
+      const readFile = attachment =>
+        new Promise((resolve, reject) => {
+          const FR = new FileReader();
```
Comment (sachaaaaa) [inline js/background.js]: Sorry for the nit, but `FR` doesn't look like a great style match with the rest of the code.
`fileReader` would be more consistent?

Comment (msgmaxim) [reply js/background.js]: This was copy pasted from `file_input_view.js`, where they already used `FR`, so I left it unchanged *for consistency* ;) I changed to `fileReader` as you suggested.

**Review by neuroscr — APPROVED** [review]

File: libtextsecure/sendmessage.js [hunk 6]
```diff
@@ -217,7 +222,10 @@ MessageSender.prototype = {
       attachmentData = result.ciphertext;
     }
 
-    const result = await server.putAttachment(attachmentData);
+    const result = isAvatar
+      ? await server.putAvatar(attachmentData)
+      : await server.putAttachment(attachmentData);
```
Comment (sachaaaaa) [inline libtextsecure/sendmessage.js]: Would `await server.putAvatar(attachmentData)` possibly only return `{ url }` but not an id? is that ok?

Comment (neuroscr) [reply libtextsecure/sendmessage.js]: Looks like that would be fine, he's just checking to make sure both don't return a falsish value (failed uploading).

Comment (msgmaxim) [reply libtextsecure/sendmessage.js]: I'm not using the `id` anywhere for avatars (should I?) and the rest of the code still works, so I omitted the field.

Comment (sachaaaaa) [reply libtextsecure/sendmessage.js]: My question was related to those lines below:
```js
const { url, id } = result;
proto.id = id;
```
We're setting `undefined` to `proto.id` but we're not using `proto.id` right?

Comment (msgmaxim) [reply libtextsecure/sendmessage.js]: Correct, `proto.id` will be `undefined`.

**Review by sachaaaaa — APPROVED** [review]
Looks good!

**Pull request merged**